### PR TITLE
Implement memory.grow()

### PR DESCRIPTION
**Pull request by Superice666**
Implement memory.grow() to allow expanding memories from java.

### CODE REVIEW

File: src/main/java/io/github/kawamuray/wasmtime/Memory.java [hunk 1]
```diff
@@ -20,6 +20,8 @@ public Memory(Store store, MemoryType type) {
 
     public native int size();
 
+    public native int grow(int page);
```
Comment (kawamuray) [inline src/main/java/io/github/kawamuray/wasmtime/Memory.java:23]: I think the argument name should be one of `pages` (plural), `delta` (by rust's API) or `delta_pages` (most understandable?)

File: src/test/java/io/github/kawamuray/wasmtime/MemoryTest.java [hunk 2]
```diff
@@ -5,8 +5,11 @@
 
 import java.nio.ByteBuffer;
 import java.util.Collections;
+import java.util.logging.Level;
```
Comment (kawamuray) [inline src/test/java/io/github/kawamuray/wasmtime/MemoryTest.java:8]: Those imports are not necessary apparently?

File: wasmtime-jni/src/io_github_kawamuray_wasmtime_Memory/imp.rs [hunk 3]
```diff
@@ -47,4 +47,10 @@ impl<'a> JniMemory<'a> for JniMemoryImpl {
         let mem = interop::get_inner::<Memory>(&env, this)?;
         Ok(mem.size() as jint)
     }
+
+    fn grow(env: &JNIEnv, this: JObject, page: jint) -> Result<jint, Self::Error> {
+        let mem = interop::get_inner::<Memory>(&env, this)?;
+        let re = mem.grow(page as u32)? as i32;
```
Comment (kawamuray) [inline wasmtime-jni/src/io_github_kawamuray_wasmtime_Memory/imp.rs:53]: 1. `as jint` instead.
2. Just `Ok(mem.grow(page as u32)? as jint)` should be fine, or please give it descriptive name (like `prev_pages`) if you intend to use this temporary variable for that purpose.

**Review by kawamuray — CHANGES_REQUESTED** [review]
Thanks for another PR! Just few minor points, overall looks good.

**Comment by Superice666** [comment]
@kawamuray  OK, done!

**Comment by Superice666** [comment]
@kawamuray As `grow` method being implemented, the memory can grow really big and its size will be larger than `INT_MAX`, which is the limit of java.nio.ByteBuffer. We will not able to read some parts of a memory then. Any solutions?

**Comment by kawamuray** [comment]
> which is the index limit of java.nio.ByteBuffer

oh, that's really a good point. The only use cases I've ever tested didn't created more than 2GiB memory so I didn't realized that issue.

I think we have 2 possible solutions for that:

1. Stop using `ByteBuffer` interface and introduce own "buffer" type that directly maps to host memory address space.
2. Add another method `ByteBuffer buffer(long offset)` or `ByteBuffer buffer(long offsetPages)` (offset by address or by number of pages) that computes base address by the given offset.

I think the first option's pros is that we can use a single "buffer" object to access entire memory space, but it also has a big downside which that we will be disabled to use java APIs that often uses `java.nio.ByteBuffer` as the type to take a chunk of data. Also, mimicing whole APIs provided by `ByteBuffer` by ourselves is a big deal.
On the other hand, the 2nd option has a little downside that users may have to manage multiple `ByteBuffer` instances to access the range of memory that they want to access, but we can keep using `ByteBuffer` which is very familiar API for many java developers and is compatible with many useful APIs.
Creating a subclass of `ByteBuffer` which supports long version methods like `position(long)` isn't an option since it is nearly impossible to create its subclass by design.

So how about this:

1. Add a new method `ByteBuffer buffer(long offsetPages)` that uses offset slice to call `env.new_direct_byte_buffer()` like `env.new_direct_byte_buffer(&mut ptr[offset..])`
2. User code calls it with offset to obtain the space they want to access. For example, if they want to access last 500MiB portion out of 3.0GiB memory, they can do `buffer = memory.buffer(2684354560 / WASM_PAGE_SIZE(=64k))`.

**Comment by Superice666** [comment]
@kawamuray Could you please merge this pull request? I think we can talk about the two solutions in another issue.

**Comment by kawamuray** [comment]
>  I think we can talk about the two solutions in another issue

ok let's do that.

**Review by kawamuray — APPROVED** [review]

**Comment by kawamuray** [comment]
LGTM, thanks!

**Pull request merged**